### ISSUE TITLE
TorchScript model export used in Andoid app Tensor data float array problem

### ISSUE THREAD

**Issue by atanasko** [ultralytics/ultralytics#773]
### Discussed in https://github.com/ultralytics/ultralytics/discussions/737

<sup>Originally posted by **atanasko** January 31, 2023</sup>
I try to use the following code

`
Bitmap resizedBitmap = Bitmap.createScaledBitmap(mBitmap, PrePostProcessor.mInputWidth, PrePostProcessor.mInputHeight, true);
`

`
final Tensor inputTensor = TensorImageUtils.bitmapToFloat32Tensor(resizedBitmap, PrePostProcessor.NO_MEAN_RGB, PrePostProcessor.NO_STD_RGB);
`

`
 IValue inputs = IValue.from(inputTensor);
`

`
Tensor outputTensor = mModule.forward(inputs).toTensor();
`

`
 final float[] outputs = outputTensor.getDataAsFloatArray();
`

to produce the output array to be able to draw the bboxes, but my result array does not look that contains the correct data. From the doc 

> The output of an object detector is a set of bounding boxes that enclose the objects in the image, along with class labels and confidence scores for each box.

but the output array look like it contains only bboxes (not sure that they are correct).

Model was custom trained model and exported with 

`
yolo export model="./runs/detect/train16/weights/best.pt" format=torchscript
`

When I test the model with 

`
yolo detect predict model="./runs/detect/train16/weights/best.pt" source="rocket_launch.mp4" show=True
`

all works fine.

Android Pytorch dependencies

`
implementation 'org.pytorch:pytorch_android:1.12'
`
`
implementation 'org.pytorch:pytorch_android_torchvision:1.12'
`

Any advice how to check is export script is valid?

I try to test the model like

`
yolo detect predict model="./runs/detect/train16/weights/best.torchscript" source="rocket_launch.mp4" show=True
`

and that look good to me, but when I try to use the model in the Android app when I inspect the output Tensor it have shape [1, 8, 8400] and I have 4 classes, and I would expect that the output float array to contain (x, y, w, h) + 4 class score, but when I inspect the array it look to me like the first half of the array are bboxes and the second half of the array the scores ex.

`
[33348]  bb : 503.1715, 507.16373, 505.32724, 505.98315 [66948]  score: 1.365644E-4, 0.0014117783, 0.45453435, 0.13102075
`
`
[33368]  bb : 498.7717, 505.04846, 503.06573, 499.7424 [66968]  score: 3.189207E-4, 0.008651144, 0.64653474, 0.36695477
`
`
[33388]  bb : 496.12158, 501.20477, 499.07532, 500.95465 [66988]  score: 4.076895E-4, 0.0055294246, 0.4945656, 0.24356054
`
`
[33408]  bb : 504.01505, 513.1348, 512.05286, 516.615 [67008]  score: 3.494384E-4, 0.0032577545, 0.33719948, 0.14475583
`
`
[33428]  bb : 392.64484, 466.43222, 482.61945, 487.92276 [67028]  score: 5.5740704E-4, 0.013573675, 0.5446357, 0.36364305
`
`
[33448]  bb : 379.116, 388.1064, 403.084, 421.47614 [67048]  score: 3.7921825E-4, 0.013157838, 0.57693213, 0.2928344
`
`
[33468]  bb : 371.716, 382.32294, 369.3178, 368.37723 [67068]  score: 8.664998E-5, 0.0011737087, 0.3274485, 0.037356436
`

What am I missing here?

**Comment by atanasko** [comment]
Anyone to help where can I find a doc or advice how to decode the output Tensor data when I use TorchScript export format? I can not understand when I use custom dataset, how data are stored in output float array get from Tensor object. I assume Tensor [1, 8, 8400] should contain 8400 tuples (x, y, w, h, p1, p2, p3, p4) p - prediction for particular class (in my case 4 classes).

**Comment by JustBrandonLim** [comment]
@atanasko I am also facing the same issue, but w ONNX and OpenCvSharp (C#).

```
ex: 3726, CX: 193, CY: 185, W: 182.28397, H: 180.75195, P1: 180.89438, P2: 178.58307, P3: 171.65671, P4: 180.5806, P5: 177.65756
Index: 3727, CX: 174, CY: 174, W: 175.52582, H: 177.57544, P1: 180.21918, P2: 183.57175, P3: 186.23444, P4: 180.48831, P5: 179.63962
Index: 3728, CX: 179, CY: 180, W: 181.82166, H: 179.7857, P1: 179.78717, P2: 177.02997, P3: 173.45203, P4: 173.2619, P5: 164.46594
Index: 3729, CX: 160, CY: 170, W: 174.30688, H: 174.72931, P1: 174.46042, P2: 176.79614, P3: 178.6618, P4: 181.17853, P5: 182.99365
Index: 3730, CX: 175, CY: 179, W: 180.08313, H: 177.79172, P1: 177.44531, P2: 173.33774, P3: 171.97931, P4: 170.00128, P5: 170.51865
Index: 3731, CX: 169, CY: 158, W: 154.48352, H: 162.65912, P1: 166.51959, P2: 167.73083, P3: 171.58667, P4: 175.11212, P5: 176.28204
Index: 3732, CX: 177, CY: 180, W: 172.95856, H: 177.93225, P1: 180.724, P2: 178.60138, P3: 178.31616, P4: 175.14832, P5: 173.1488
Index: 3733, CX: 170, CY: 169, W: 172.59653, H: 2.784831E-11, P1: 2.3250566E-11, P2: 8.385248E-12, P3: 1.6461072E-11, P4: 1.10880515E-11, P5: 6.073912E-11
Index: 3734, CX: 0, CY: 0, W: 2.2706174E-10, H: 1.5078391E-10, P1: 9.995086E-11, P2: 8.5783415E-11, P3: 9.7324454E-11, P4: 1.05156175E-10, P5: 1.2258272E-10
Index: 3735, CX: 0, CY: 0, W: 1.4782096E-10, H: 1.5669539E-10, P1: 1.6537646E-10, P2: 1.6167313E-10, P3: 1.5075889E-10, P4: 1.5706193E-10, P5: 1.6889541E-10
Index: 3736, CX: 0, CY: 0, W: 2.1849397E-10, H: 2.2252393E-10, P1: 2.2807307E-10, P2: 2.4019672E-10, P3: 2.5858168E-10, P4: 2.8148123E-10, P5: 3.1658412E-10
Index: 3737, CX: 0, CY: 0, W: 4.0424897E-10, H: 4.1778503E-10, P1: 4.4602785E-10, P2: 4.567166E-10, P3: 4.6881715E-10, P4: 5.1170523E-10, P5: 5.6073585E-10
```

I have 5 classes in my custom trained model.

**Comment by atanasko** [comment]
@JustBrandonLim I think I figure it out. So YOLOv8 model generate output Tensor [1, 8, 8400] (in my case). When data float array is generated from this Tensor, because of the Tensor shape,  indices of the first bbox and scores for the first bbox in the array are:

x   ->  0 * 8400
y   ->  1 * 8400
w  ->  2 * 8400
h  ->   3 * 8400
p1 -> 4 * 8400
p2 -> 5 * 8400
p3 -> 6 * 8400
p4 -> 7 * 8400

I modify YOLOv5 Android app, and you can find the modified Android app at [ObjectDetection](https://github.com/atanasko/android-demo-app/tree/master/ObjectDetection)

Hope this helps

**Comment by larrywal-express** [comment]
Good job. Having test the Object Detection app, it worked very good. How to incorporate fps into the app?

**Comment by larrywal-express** [comment]
@atanasko the display of detection time in fps/ms

**Comment by atanasko** [comment]
@larrywal-express Is this that you want to get?

`
    private Deque frameTimestamps = new ArrayDeque <Long>(5);
    private int frameRateWindow = 8;

  // Keep track of frames analyzed
            long currentTime = SystemClock.elapsedRealtime();
            frameTimestamps.push(currentTime);

            // Compute the FPS using a moving average
            while (frameTimestamps.size() >= frameRateWindow)
                frameTimestamps.removeLast();
            Long first = (Long) frameTimestamps.peekFirst();
            Long last = (Long) frameTimestamps.peekLast();
            Long timestampFirst = first != null ? first : currentTime;
            Long timestampLast = last != null ? last : currentTime;
            double framesPerSecond = 1.0 / ((timestampFirst - timestampLast) /
                    (double) frameTimestamps.size()) * 1000.0;

            System.out.println("----------------------------------" + framesPerSecond);
`

inside AbstractCameraXActivity#setupCameraX#ImageAnalysis.Analyzer

**Comment by larrywal-express** [comment]
@atanasko thanks for your quick response. I added the provided code into the AbstractCameraXActivity.java as suggested. However, the detection time is not displaying on the phone screen. Can you provide me the complete code of AbstractCameraXActivity.java. Thank you.

**Comment by atanasko** [comment]
@larrywal-express I add this for console logging as example. I did not update UI with the info. I can send the complete class but it's code already available plus the one in the comment.

**Comment by larrywal-express** [comment]
Okay. you can send me the complete code. Thanks

**Comment by atanasko** [comment]
@larrywal-express you can find the modified code [here](https://github.com/atanasko/android-demo-app/blob/fps/ObjectDetection/app/src/main/java/org/pytorch/demo/objectdetection/AbstractCameraXActivity.java)

**Comment by larrywal-express** [comment]
@atanasko Thanks. it worked... but the detection time is not display for image or select or live... just like the attached image sample
![fps](https://user-images.githubusercontent.com/62125055/225174676-aed2d1bb-70cb-40b7-a76f-130d412c478e.jpg)

**Comment by larrywal-express** [comment]
@atanasko This is another sample displaying the detection in real-time
![Screenshot_20230315_085641](https://user-images.githubusercontent.com/62125055/225175972-4a72b5ef-0664-4177-8b70-42cfc53a8aa0.jpg)

**Comment by atanasko** [comment]
hi @larrywal-express, can you please try the code from fps branch in live mode?

**Comment by larrywal-express** [comment]
I tried... but no display

**Comment by atanasko** [comment]
@larrywal-express are you using the "fps" branch?

![rocket_detect](https://user-images.githubusercontent.com/3932183/225309690-25495544-7a09-4da8-8688-c2f8b9b381a2.jpg)

**Comment by larrywal-express** [comment]
yes.. Am using the fps branch

**Comment by atanasko** [comment]
@larrywal-express and also "Live" mode right? I do the change for video only

**Comment by larrywal-express** [comment]
yes... only for live... I will download again and test it

**Comment by larrywal-express** [comment]
Good.. it is working fine now... Thanks.

**Comment by pderrenger** [comment]
@atanasko it's great to see that your issue has been resolved! If you have any further questions or need assistance with anything else, feel free to ask. We're here to help!
